### ISSUE TITLE
Why retrieve a list of tuners instead of just one tuner? 

### ISSUE THREAD

**Issue by tidoust** [w3c/tvcontrol-api#4]
The main entry-point of the TV Control API is the retrieval of a list of tuners through the `getTuners` function.

I would have assumed that, leaving recorders aside for a moment, most devices ship with only one tuner, or at least one tuner of given kind (radio/tv). What are the use cases that drive the need to expose a list of tuners? Are there common products that feature multiple tuners?

What can a Web app do with the list of tuners it gets from `getTuners`? It cannot easily ask the user to select a tuner for instance, because all it has to describe the tuner is an ID which most probably won't mean anything to the user.

I wonder whether that function should not be replaced by a `getTuner` function that could possibly take a filtering rule as parameter and would return one `Tuner` instance. The user agent is likely going to prompt the user for permission to access the tuner when this call is issued in any case, so it could also use that opportunity to ask the user to select the tuner to expose to the Web app if different tuners are available.

Note that I left tuners used for recordings aside because the spec treats them separately as well.

**Comment by stevem-tw** [comment]
Some TVs from major manufacturers will contain more than one type of tuner (e.g. cable and terrestrial, or satellite and terrestrial), and I have seen TVs and STBs with three types of tuner (cable, satellite and terrestrial).  We can't assume that all tuners will be able to receive the same set of channels, or that all of them will have an input connected.

The list of tuners seems to be the only entry point to getting a channel list or EPG data, so unless that is made available in another way then a web app has no choice but to use `getTuners`.

**Comment by tidoust** [comment]
Hi @stevem-tw,

> Some TVs from major manufacturers will contain more than one type of tuner 

Understood, thanks! That answers the first part of my question: there are products out there with multiple tuners. I'm still not clear what a Web app can do with such a list in practice, without meaningful information to display to the user. Is it going to parse the list of channels that each tuner exposes to select the tuner it wants to use? How would it know?

> The list of tuners seems to be the only entry point to getting a channel list or EPG data, so unless that is made available in another way then a web app has no choice but to use  getTuners .

Right, note that I was not suggesting to only drop the function, but also to replace it with a `getTuner` function that would return only one `TVTuner` instance. The `TVTuner` instance is the one that provides access to a list of channels and EPG data (through another `TVSource` level).

**Comment by stevem-tw** [comment]
>  I'm still not clear what a Web app can do with such a list in practice, without meaningful information to display to the user. Is it going to parse the list of channels that each tuner exposes to select the tuner it wants to use? How would it know?

The Web app will have some meaningful information, in that it has the name and logical channel number of each channel.  It may also choose to combine this with information from another source (e.g. EPG data that it's retrieved via XHR).

> Right, note that I was not suggesting to only drop the function, but also to replace it with a getTuner function that would return only one TVTuner instance. The TVTuner instance is the one that provides access to a list of channels and EPG data (through another TVSource level).

At the moment, I'm not 100% clear on the difference between a `TVTuner` and a `TVSource`.  I'm probably missing something but it's not clear which of the two represents an actual, physical tuner.  The spec as it is written implies that a `TVSource` object actually represents a physical tuner (since this is what a web app will use to tune to a specific channel, and which is associated with a specific delivery system and standard (e.g. DVB-S, ATSC or another).  

If this is the case, I'm not sure what it means to have multiple 'TVTuner' objects - the only case I could think of would be a PC which has multiple TV tuner devices attached, where each device has a separate `TVTuner` instance. For a typical STB or TV, I would expect that the user agent would have a single `TVTuner` instance, which exposed a `TVSource` for each physical tuner in the device. This is an implementation question, though, and so my expectation may not be correct. 

Depending on how correct my assumptions above are, it may be that it would be more useful to have a function that gets a `TVSource' of a specific type.

As I said, though, I could easily be missing something important, so any clarifications are welcome if I have missed anything!

**Comment by tidoust** [comment]
> At the moment, I'm not 100% clear on the difference between a  TVTuner  and a  TVSource .

Indeed, I think that's the core of the issue. I suppose these 2 levels match some hardware reality, but Web app developers are unlikely going to get that and this seems to make the API unnecessarily complex.

> Depending on how correct my assumptions above are, it may be that it would be more useful to have a function that gets a `TVSource' of a specific type.

I agree but we may both be missing something important. Does anyone know what led to the current design that separates `TVTuner` and `TVSource`?

**Comment by AMorgaut** [comment]
**About having multiple Tuners**

We are actually working on a STB which includes 4 tuners and we know some versions includes at least up to 8 tuners.

Each tuner can be given one specific tasks such as:
- perform a channel scan
- play a channel stream, as the main Live, or as a PiP live
- record a channel 
- record & play a channel at the same time

It might not be a generic rule, but still a common one, even if it is technically possible, we could not get video streams of multiple channels at the same time from a single tuner, so multiple tuner meant here the possibility to perform multiple tasks of this kind in parallel 

ex. Playing live on 1 channel + Live PiP on another + recording 2 other Live channels

If you check on the Web, many people look for solutions to play multiple streams from a single tuner, and when they succeeded, it was a big Houra because not simple at all. Even when they did it, it happens that they sometimes already had multiple channels but wanted to extend at least a little the number of parallel streams (ex: 8 streams from 4 tuners with 2 stream per tuner)

What has been the most important for us, as app developers, was to be aware if there was some available tuners at a due moment to perform a required task without stopping already running or scheduled ones (background channel scan for updates, detection of parallel record conflict, ...)

Then, going deeper and at least knowing on which Channel is tuned a Tuner may be interesting to anticipate the expected delay before at least one tuner is ready to stream an intended Channel. 

I'm not sure that it is necessary to give the possibility to the app developer the possibility to choose himself which tuner should tune to which channel. The platform implementer might want to be able to perform its own optimisations depending on how its hardware behave. Still, things that might be important to take into account:
- as said tune a tuner to a specific channel can take time
- enable / disable subtitle on that channel can take time
- switch audio or subtitle on that channel can take time
- changing the channel of tuner usually means breaking its PLTV (Pause Live TV) buffer

It's hard for the platform to anticipate which actions will be successively sent to a tuner. Usually you disable audio & subtitles on streams used for PiP, and sometimes even the PLTV buffer. So you might have the notion of sessions to abstract tuner actions.

**TVTuner vs TVSource**

Some tuners can sometimes read broadcast stream from different kind of sources. So instead of embedding a DVB Terrestrial tuner + a DVB Cable Tuner + a ATSC (Cable /Terrestrial) Tuner so your SmartTV or STB can be used in more countries on both Terrestrial & Cable signals, a constructor might only expose a single Tuner which will switch its reception mode depending on the available TVSources. 

You should hardly have in runtime concurrent DVB & ATSC TVSources on a TVTuner, but I think that concurrent DVB-T & DVB-C sources is a pretty common use cases. As an example an operator client had a limitation on the number of channels it could push to its end-users. So it wanted to limit cable channels to only those that were not available through Terrestrial signal and merge both of the channel lists. So having a Tuner that can switch between the 2 TVSources is here pretty useful.

**Comment by stevem-tw** [comment]
@AMorgaut thanks for the feedback.  Most set-top boxes today have multiple tuners, and so this is definitely an area that needs careful handling.  Experience has shown me that, as you say, trying to be too clever with tuner usage leads to unpredictable behaviour and confusion for end users (since you can successfully re-use tuners in some cases but not others, and the interaction between different use cases is complex). This is why I tend to prefer an approach which makes tuner management more explicit, like the API currently does: other standards that I've been involved with have taken similar approaches for the same reason. There is still scope for some optimization by the hardware implementer if they want to do so, but it gives predictability to the developer and end user.

> Some tuners can sometimes read broadcast stream from different kind of sources. So instead of embedding a DVB Terrestrial tuner + a DVB Cable Tuner + a ATSC (Cable /Terrestrial) Tuner so your SmartTV or STB can be used in more countries on both Terrestrial & Cable signals, a constructor might only expose a single Tuner which will switch its reception mode depending on the available TVSources.

You're right that this is a good reason to separate TVTuner and TVSource, and it is probably worth documenting clearly this in the spec.  One of my concerns is that the relationship between the two classes is not well specified - re-reading the spec, this is implicit but isn't actually stated anywhere..  This may be something that needs to go in the introduction (or in the definitions of the TVTuner and TVSource classes) so that we have a common understanding between implementers of the API and developers using the API.

> You should hardly have in runtime concurrent DVB & ATSC TVSources on a TVTuner, but I think that concurrent DVB-T & DVB-C sources is a pretty common use cases. As an example an operator client had a limitation on the number of channels it could push to its end-users. So it wanted to limit cable channels to only those that were not available through Terrestrial signal and merge both of the channel lists. So having a Tuner that can switch between the 2 TVSources is here pretty useful.

Agreed, but I think we need to make sure that the developer can distinguish this case from the case where there are two physical tuners that are not switchable between different source types: this is where we may need some extra text to clarify this.

**Comment by AMorgaut** [comment]
> I think we need to make sure that the developer can distinguish this case from the case where there are two physical tuners that are not switchable between different source types: this is where we may need some extra text to clarify this.

I think that this is the purpose of `getSupportedSourceTypes()`  TVTuner method.
With it, the developer should be able to know exactly which Tuner can give access to which kind of TVSource even before trying to reach them. But I think `getSources()` could then benefit from an optionnal `sourceTypeFilter` param.

**Comment by tidoust** [comment]
There may be multiple discussions in one here which may be worth pursuing in separate issues, but let's see where this brings us first. I feel that I may still be missing something fundamental here. On TVTuner vs. TVSource, @stevem-tw wrote:

> > I think we need to make sure that the developer can distinguish this case from the case where there are two physical tuners that are not switchable between different source types: this is where we may need some extra text to clarify this.

@AMorgaut replied:

> I think that this is the purpose of `getSupportedSourceTypes()` TVTuner method.
> With it, the developer should be able to know exactly which Tuner can give access to which kind of TVSource even before trying to reach them. But I think `getSources()` could then benefit from an optionnal `sourceTypeFilter` param. 

Intuitively, I would prefer a tuner with multiple `TVSource` to behave as a tuner with only one `TVSource`. Other than for informational purpose (which could be returned as part of the channel's metadata), why would a Web app developer need to know that a tuner may switch between multiple sources?

**Comment by stevem-tw** [comment]
> Intuitively, I would prefer a tuner with multiple `TVSource` to behave as a tuner with only one `TVSource`. Other than for informational purpose (which could be returned as part of the channel's metadata), why would a Web app developer need to know that a tuner may switch between multiple sources?

A Web app developer may not care, but this is a decision that has a substantial impact on how you would implement the API.  At the moment, things like channel lists and logical channel numbering and channel scanning are managed on a per-source basis.  If you removed the concept of a `TVSource` then the implementation would have to provide a single channel list that is merged from all sources.  However, this would have a number of complications:
- channel numbering schemes may not be consistent across sources
- channel scanning may have different performance issues or algorithms depending on different sources
- the same channel may be present on multiple sources, which may lead to a unified channel list having several instances of the same channels.

Without specifying the behaviour in extreme detail, it's not really possible to guarantee that multiple implementations would give consistent results, and TV and STB manufacturers would probably object to that, given that they already have implementations of this functionality that they would be very reluctant to modify.

**Comment by paulhiggs** [comment]
From my recollections of the TV Control API community group, the TVTuners provided a dynamic list of tuning resources that are available in the device (note that we had some requirements that tuners could be added or removed from the system during run-time perhaps through USB interface). The sequence<TVSource> in each TVTuner designates what delivery types the TVTuner supports (perhaps a tuning resource could simultaneously support DVB-C and DVB-C2. It may be that in 99% of devices, TVTuner.getSources() yields a sequence of just one element.

**Comment by chrisn** [comment]
At the TPAC F2F meeting we agreed to make the API more source-centric. See the discussion [here](https://www.w3.org/2016/09/20-tvapi-minutes.html#item05).

A problem with the existing API we discussed is that the relationship between tuners, sources and stream decoding resources is unclear, and that we should aim to simplify the API for application developers.

We sketched a possible source-centric API, copied here from the minutes:

``` javascript
interface TVManager : EventTarget {
  Promise<sequence<TVSource>> getSources();
};

interface TVSource : EventTarget {
  Promise<TVTuner> tuneToChannel(TVChannel? channel);
  Promise<sequence<TVChannel>> getChannels();
  int getNumberOfTuners();
};

interface TVTuner : EventTarget {
  attribute TVMediaSTream stream;
  Promise<> stop();
  Promise<> setChannel(TVChannel channel);
};

// Usage example
TVManager.getSources().then(sources => {
  let source = sources[0];
  source.getChannels().then(channels => {
    let channel = channels[0];

    // This would tune to the "current" channel
    source.tuneToChannel().then(tuner1 => {
      video.srcObject = tuner.stream;
    });

    // This would tune in to a specific channel
    source.tuneToChannel(channel).then(tuner2 => {
      video.srcObject = tuner.stream;
    })
    .catch(err => { 
      console.log('Oh no! No way to get a tuner');
      tuner1.stop();
    });
  });
});
```

There may be a need for applications to be able to select a particular tuner based on capabilities:

> This doesn't work properly, because tuners are not interchangeable. A tuner may support HD, while another one may only support SD for instance.
> 
> But then, there may be a capability issue, if one stream is used for picture in picture, using SD, and the main stream is used for HD. If the HD tuner is locked for PiP, there is a problem.
> 
> Another solution could be to expose a getTuners method on TVSource:

``` javascript
interface TVSource : EventTarget {
  Promise<sequence<TVTuner>> getTuners();
  Promise<sequence<TVChannel>> getChannels();
  int getNumberOfTuners();
};
```

> The takeaway from the discussion is that the group agrees to move to a source-centric approach for the API. The issue as to whether we expose the list of tuners or dynamically map channels to tuners remains an open issue at this stage.

Further comments on this API suggestion are welcome (and encouraged!)

**Comment by JPEvain** [comment]
Hi all,

I am probably joining the discussion late but I am starting to dive into this as my problem is to look at a better use of metadata for selecting a programme of interest to be eventually displayed.

Reading the specification, independently from implementation details (i.e. what function should do what), I can live with a TVManager that identifies available TVTuners (yes, tv sets on the market do have more than one tuner) as physical devices, through which information can be collected to list tvsources/channels.

Question: do we mean using inline metadata from embedded service information? 

Then either a tvsource/channel is selected and the mediaStream in consumed displaying the programme available at that time, or the inline metadata such as present/following data allows selecting a tvsource/channel for the particular programme it offers.

Question: I do not see how the current APIs would exploit the data to select a tvsource/channel for its content/programme.

Actually, in this context, I am more confused with "tvsources vs. channels" than "tvtuner vs tvsource".

General comment: we have clear focus on TV but these could be Radio as well

**Comment by JPEvain** [comment]
I forgot to mention that I have seen mention of the EPG. In the current context, an EPG could be reconstructed from the inline present/following SI information from the different tvsource/channels.

The other option would be to allow linking EPG items from an online (e.g. OTT) application to tune to the right tuner/source/channel/programme

**Comment by JPEvain** [comment]
It's going to be slow progress through the spec, logs and github :--)

Ok I found the getProgramOptions that leads to a program description (most likely from inline data?).

**Comment by JPEvain** [comment]
I don't know the details of the genesis of the specification.

There is definitely potential confusion around some of the key concepts like source, channel or else.

Is there somewhere a mapping to dvb and atsc termonilogie (as as tv is concerned). MIght be interesting to see how the conpcet of network, bouquet, service, event, etc. would map to source / channel / event / programme or even mediaStream.

What do you think? Mayne it exists?

**Comment by JPEvain** [comment]
I have not been working on this for quite some, so I had to do some digging...

Example of mapping to DVB-SI

DVB Network -> TVControl Tuner
DVB Multiplexes - > TVControl sources
DVB Services -> TVControl channels
DVB Bouquet (group of services) -> null
DVB Event -> TVControl program
DVB Components - > TVControl MediaStreams (video, audio ..). or a program or mediastream fed through the tuner?)

**Comment by JPEvain** [comment]
Continuing on this...
Step 1 - identify tuners
Step 2 - for each tuner, scanning frequencies and identify sources/Multiplexes
Step 3 - Identify channels/services
Step 4 - Identify programs/events
Step 5 - select and play mediaStreams

Q1 - should the information from step 1,2 and 3 be also available from the TV setup?
Q2 - we need to clarify media streams. I think I have read somewhere that a stream might be interpreted as a stream when tuning on a frequency. However, better reserve stream to audio video and other streams needed for rendering via the <video> or <audio> attributes.
Q3 - have we defined the format of the locator in  <video> or <audio>?

**Comment by stevem-tw** [comment]
@JPEvain I don't think that there is a good mapping between the objects in the API at the moment and DVB-SI (apart from DVB Service -> TVChannel and DVB Event -> TVProgram).

I think we need to clarify the definitions used in the API - I'm not sure that the descriptions of those objects are precise enough at the moment for us to have a shared understanding of the API.  Based on the interface posted by @chrisn  above that was discussed at TPAC, here are the descriptions that I've been working with in my head:
- **TVSource**:  a logical source of TV/Radio channels.  This may represent a physical tuner or a "virtual" tuner for IP multicast.  The TVSource exposes the list of TVChannel objects available through that source.  For devices with more than one type of tuner (e.g. cable and terrestrial), each tuner type may be represented by a different TVSource.  A TVSource is **not** a physical tuner - it is a way of discovering channels, getting the corresponding TVChannel objects and obtaining the resources to present them. Because there is not a one-to-one mapping between a TVSource and a tuner, more than one channel from a given TVSource may be presented at the same time if the device has enough resources.
- **TVChannel**: represents a single TV or radio channel and provides metadata about it.  Each TVChannel object represents a single channel that is available through one TVSource.  If the same channel (e.g. VH1) is available through two different sources (e.g. over cable and terrestrial) then each TVSource will return a different TVChannel object.
- **TVMediaStream**: represents the stream data for all components within a channel.  This can be passed to an HTML5 media element to play the channel.  Obtaining a TVMediaStream object means that all of the resources needed to receive and present that channel have been allocated. I'm not sure how closely this aligns with your view, @JPEvain.
- **TVTuner**: I'm really not sure about this.  The obvious solution is for this to represent a physical tuner, where a TVSource object may have one or more TVTuner instances associated with it (one per physical tuner).  However, this has a number of problems.  First, it doesn't accurately represent modern tuner hardware that supports full-band capture, where a single tuner can act as the source for multiple streams and you're no longer limited to one transport stream per physical tuner.  Second, it also fails to represent cases such as channels delivered via IP. Third, it doesn't represent all of the other resources that are needed to decode and present a stream.  While we could change the semantics of this object to represent "the end to end decoder chain", I'm not sure that's appropriate.  That would be a way of exposing what we discussed at TPAC (and which @paulhiggs mentioned elsewhere) about exposing the capabilities of a device in terms of the number and type of services that it can decode simultaneously.

Thoughts and comments on this are welcome.

**Comment by tidoust** [comment]
That seems like a very good starting point for the terminology section of the spec, @stevem-tw :)

One comment about the `TVMediaStream` definition: it seems to imply that `TVMediaStream` are attached to `TVChannel`. That's what I proposed initially at TPAC, but we also discussed the possibility to reuse the same stream across channels, which seems possible, similar to the way `getUserMedia` operates, and also aligned with how a WebRTC peer "sees" an incoming media stream: even if the stream changes, e.g. because the user switched to another camera, I think the peer still receives the same stream.

In this vision, a `TVSource` would produce a `TVMediaStream` that can be further "constrained" afterwards (typically by allowing to switch to another channel). A `TVMediaStream` represents "the stream data for all components within a channel" but may also represent the stream data of another channel later on.

This also seems to be the approach the TAG suggests in issue #20 as well (the TAG's comment uses `TVTuner` but I think that's only because it is the main entry point right now).

One thing that makes this a bit more complex is that a `TVSource` can produce more than one `TVMediaStream` in some cases. This is where some of notion of `TVTuner`, defined as a `TVMediaStream` **wrapper and controller**,  could perhaps be useful, although it may not be really needed in practice. Now, that probably matches "the end to end decoder chain" that you allude to, so I may just be paraphrasing your comment here ;)

One of the arguments mentioned at TPAC in favor of keeping a `TVTuner` class was to provide an extension point to the API. I feel that this can be achieved in the definition of capabilities and constraints associated with a `TVSource` though.

All in all, I guess I would start the source-centric design **without** the `TVTuner` interface... and only re-introduce it if design looks clumsy, if things break or if requirements cannot be satisfied.

**Comment by JPEvain** [comment]
The list of definitions is what we want to achieve to make sure we are all  speaking the same language but there might be more work to be done. I'll share my comments soon.

As I mentionned earlier in the thread, I believe mapping to DVB and ATSC is important in order to interoperate with these systems and the information they carry.

I have been going back to the list of use cases to derive a few typical configurations. I was pleased to see that the use cases cover the configurations I had in mind when thinking about the definitions and associated data and some basic implementations. These configurations are important to remind us why we want to use the API for. Maybe we should open a separate issue to discuss / list configurations...  
1. A device (not a TVSet) is equipped with a tuner and the proposal is to control/use this tuner to access content via an application
2. An (e.g. OTT) application like an EPG is used to select content of interest. The proposal is to play the content on the TVSet (with its own setup process),  although it could equally be played on the application mentioned above.

Is there anything else important missing? If not, we can look at the definitions again to:
- refine what we associate with each term / object accessible through the API
- verify the mapping between the data provided in line, which will be used to discover and access content through the API, and the data provided in applications like an EPG or why not a search engine.

I'll look at the definitions again with these two configurations in mind and comment bearing in mind my original attempt to map to DVB (I know where to find DVB support) and ATSC (ATSC experts please help).

**Comment by JPEvain** [comment]
Of course, manufacturers in the group are more than welcome to tell us how an application can access a tuner, retrieve (what) data, etc.

**Comment by stevem-tw** [comment]
@jpevain I agree that we need to define a mapping between the API and DVB/ATSC/others, but I'm not sure we're in a position to do that yet, until we've agreed on the overall "shape" of the API.  I can comment some on the ATSC side, although I'm more familiar with DVB.  If we broadly agree with the definitions I proposed above (except for TVMediaStream; see below) then we can look at that mapping in more detail.

@tidoust You're correct that my proposal makes the assumption that a TVMediaStream is associated with a channel, and this does not accurately reflect the discussions we had at TPAC.  I agree with your comments about TVMediaStream, but I'm not sure how to phrase that in terms of a definition at the moment :)

**Comment by stevem-tw** [comment]
I've tried re-phrasing the definitions from up-thread to more accurately reflect the discussions around TVMediaStream from TPAC.  I've left TVTuner out of this for now, because I don't think we're close enough to agreement on that at the moment.
- **TVSource**: a logical source of TV/Radio channels. This may represent a physical tuner or a "virtual" tuner for IP multicast. The TVSource exposes the list of TVChannel objects available through that source. For devices with more than one type of tuner (e.g. cable and terrestrial), each tuner type may be represented by a different TVSource. A TVSource is not a physical tuner - it is a way of discovering channels, getting the corresponding TVChannel objects and obtaining the resources to present them. Because there is not a one-to-one mapping between a TVSource and a tuner, more than one channel from a given TVSource may be presented at the same time if the device has enough resources.
- **TVChannel**: represents a single TV or radio channel and provides metadata about it. Each TVChannel object represents a single channel that is available through one TVSource. If the same channel (e.g. VH1) is available through two different sources (e.g. over cable and terrestrial) then each TVSource will return a different TVChannel object.
- **TVMediaStream**: represents the stream data for all components within a channel being presented. This can be passed to an HTML5 media element to play the channel. A TVMediaStream is obtained from a TVSource, and when a TVMediaStream object is created it means that all of the resources needed to receive and present that channel have been successfully allocated.  A TVMediaStream object may be re-used when selecting a different channel, to indicate that tuning and decoding resources currently used to present that TVMediaStream may be re-used for the new channel.

One thing that springs to mind is that in this model a TVMediaStream can also act as a representation of the set of resources needed to receive, decode and present the channel that it's currently being used for.  I'm not sure how or if we can use that information, but it's something that may prove useful so I thought I'd bring it to people's attention.

**Comment by tidoust** [comment]
Capturing the feedback  from @JPEvain [sent by email](https://lists.w3.org/Archives/Public/public-tvcontrol/2016Nov/0037.html) (replies to notification emails are not automatically converted to GitHub comments):

[[
Thanks for updating the definitions.
1.  Introducing the notion of IP multicast may lead to different APIs to resolve the list of TVChannels available from these different types of sources?
2. I would keep the definitions as generic as possible. The penultimate sentence looks like a repeat of what's being said in the second sentence.  I would keep the last sentence of the definition of TVSource out.
   - **TVSource**: a logical source of TV/Radio channels. This may
     represent a physical tuner or a "virtual" tuner for IP multicast. The
     TVSource exposes the list of TVChannel objects available through that
     source. For devices with more than one type of tuner (e.g. cable and
     terrestrial), each tuner type may be represented by a different
     TVSource. 
   - TBD: if associated to TVTuner: DVB Network and Multiplex, ATSC?? 
3. The definition of TVChannel -> DVB Service, ATSC??
   - **TVChannel**: represents a single TV or radio channel and provides
     metadata about it. A TVChannel objectis one among possibly several channels
     available through one TVSource. If the same channel (e.g. VH1)
     is available through two different sources (e.g. over cable and
     terrestrial) then each TVSource will return a different TVChannel
     object.
4. Mediastream (if we want to reuse definitions)
   
   A TVChannel is consumed as **TVMediaStream** composed of one or more MediaStreamTracks (audio, video, data, etc. -> DVB Components, ATSC??) This can be passed to an HTML5 media
   element to play the channel. A TVMediaStream is obtained from a
   TVSource, and when a TVMediaStream object is created it means that all
   of the resources needed to receive and present that channel have been
   successfully allocated. 
   
   To be rediscussed when we have progressed on the API?
   [[[[A TVMediaStream object may be re-used when
   selecting a different channel, to indicate that tuning and decoding
   resources currently used to present that TVMediaStream may be re-used
   for the new channel.]]]] 
5. TVProgram -> DVB Event -> ATSC??
   A TVProgram is an editorial object available at a given time on a given TVChannel
   ]]

**Comment by stevem-tw** [comment]
@JPEvain thanks for your comments

> Introducing the notion of IP multicast may lead to different APIs to resolve the list of TVChannels available from these different types of sources?

I would hope that the API is the same in each case.  We have getChannels() on the TVSource, although I do wonder whether this per-source separation of channels is really a good thing: many devices only present a unified channel list to the user, regardless of the source.  I'm not sure if that's something we need to consider now, or leave it until we've defined the basic shape of the API

> I would keep the definitions as generic as possible. The penultimate sentence looks like a repeat of what's being said in the second sentence. I would keep the last sentence of the definition of TVSource out.

Re-reading this, I agree.

> The definition of TVChannel -> DVB Service, ATSC??

It would be an ATSC Channel

> Mediastream (if we want to reuse definitions)

Yes, I think the area you highlight probably does need further discussion, although what is there is based on discussions from TPAC and is similar to the WebRTC model.

> TVProgram -> DVB Event -> ATSC??

It would be an ATSC Event.

**Comment by stevem-tw** [comment]
Foe @JPEvain 's feedback on MPEG transport streams being common to IP broadcast, given the move to MMT or multicast ABR in some regions I don't think we can assume we will be using MPEG transport stream.  While I think we need to define a mapping to MPEG-2 TS for some aspects of the API, I would not want to limit it to just supporting MPEG-2 TS.

**Comment by tidoust** [comment]
Email sent to the Media Capture TF of the Device and Sensors WG and WebRTC WG to gather feedback:
https://lists.w3.org/Archives/Public/public-media-capture/2016Dec/0038.html

**Comment by davisrc** [comment]
![image](https://cloud.githubusercontent.com/assets/8480450/21598071/99cf9fd8-d123-11e6-87d1-a9a48c9b4bbc.png)
![image](https://cloud.githubusercontent.com/assets/8480450/21598074/a227e69a-d123-11e6-9f5e-b161503c482d.png)

i am not sure where to put this but i may have put it in a couple of spots..

Here is a concept that I was working on that involves available resources.  The resources would be any available source that would share a common constraint such as only one of them in the group would be available at a time.  This also shows the concept of using “players” instead of tuners.  Players only decode and play to a specific output zone.  This would allow then an external player to be used and to use a resource from the system.  How the system defined players interact with the system would be completely up to the system to define and implement.  You would use setMedia (or something similar) to assign a specific source of a resource to a specific player.  I know, of course, this is kind of crude and needs some refinement, discussion, and contributions, but I think this would be a good way to make the api not dependent on the underlying technology.

**Comment by stevem-tw** [comment]
Following on from our [recent conf call](https://lists.w3.org/Archives/Public/public-tvcontrol/2017Jan/0003.html), this is an attempt to merge the different ideas we’ve been sharing on the shape of the API in an incomplete set of functions:

TVManager:
-	TVSourceCapabilities getCapabilities()
-	getSource(TVSourceConstraints constraints)
-	isSourceAvailable(TVSourceConstraints constraints)

TVSource:
-	TVSourceConstraints getConstraints()
-	TuneToChannel()
-	TuneTo()

The TVSource acts as a proxy for the various hardware and software resources needed to acquire a media stream, although the details of these resources are hidden from the app.  (Note: this partly maps on to Ryan’s concept of a “resource” in his diagram).  Apps don’t enumerate the set of available TV sources: instead, they request one which satisfies a particular set of constraints and the implementation tries to satisfy these constraints.  Constraints may include things like quality level (SD, HD, UHD for video; other quality levels for audio), delivery system (e.g. AM or FM for radio; other delivery systems for TV).
Apps can either attempt to get a source or can query whether a source that matches their constraints is available.  Note that this is different to GetUserMedia() where constraints are applied to tracks not sources, but I really don’t see a good way of making that work in our case.

When an app has a TVSource then it can use that to tune to a channel.  This gives a MediaStream which can be passed to a media element for playback.



One thing that needs thinking about here is resource management and how that is handled. As proposed in issue #3 I think that simply getting a TVSource does not allocate any scarce resources: this only happens when a source is tuned to a channel.  This does have implications for what it means to query whether a source is available, but I can expand on that if the basic approach is acceptable.

**Comment by chrisn** [comment]
In case you haven't seen it, the follow-up discussion to [Francois' email](https://lists.w3.org/Archives/Public/public-media-capture/2016Dec/0038) to the Media Capture TF is happening on the [public-media-capture](https://lists.w3.org/Archives/Public/public-media-capture/) mailing list. Please take a look at the various replies we've received. Your contributions to the discussion there are very welcome!

**Comment by chrisn** [comment]
@stevem-tw: Your new proposal looks good, but a few points came to mind:

> The TVSource acts as a proxy for the various hardware and software resources needed to acquire a media stream [...]

We've discussed a few names so far for the idea of "the set of resources dedicated to providing a stream" - TVTuner, Player, Session, and now TVSource. I think my preference is still for Session, although I'd probably name the interface TVMediaSession, which might help distinguish it from "session" as in cookies. But that said, I don't feel too strongly.

> I think that simply getting a TVSource does not allocate any scarce resources: this only happens when a source is tuned to a channel. 

Yes, I agree.

We'll also need to think about how we handle channel scanning in this new API structure. This is currently done on a per-source basis, so would we still need a TVSource interface (as currently written in the spec), and the ability to enumerate sources? (We should probably discuss this under issue #7; it may depend on whether we need to expose an API for scanning at all).

FInally, would we need the `isSourceAvailable` method? I imagine the main use case for this is to check whether a subsequent call to `getSource` will succeed, which we can handle through the returned promise. Are there other use cases I've missed?

**Comment by stevem-tw** [comment]
@chrisn Thanks for your comments

> We've discussed a few names so far for the idea of "the set of resources dedicated to providing a stream" - TVTuner, Player, Session, and now TVSource. I think my preference is still for Session, although I'd probably name the interface TVMediaSession, which might help distinguish it from "session" as in cookies. But that said, I don't feel too strongly.

I would be happy with this, partly because it gets us away from existing names and stops us making any assumptions.
> > I think that simply getting a TVSource does not allocate any scarce resources: this only happens when a source is tuned to a channel. 

> Yes, I agree.

There are some consequences of this approach, but I think we can deal with those.  If we say that we have a method or methods that tunes to a channel (which allocates resources) and a method which stops the session (which stops streaming the channel and deallocates any resources) then this is probably OK, but it is slightly different from the way that is used by the Media Capture and Streaming API.

> We'll also need to think about how we handle channel scanning in this new API structure. This is currently done on a per-source basis, so would we still need a TVSource interface (as currently written in the spec), and the ability to enumerate sources? (We should probably discuss this under issue #7; it may depend on whether we need to expose an API for scanning at all).

I'm inclined to say that we move the scanning API to the TVManager interface and say that scanning applies to the device as a whole, not a particular source.  This has the advantage of meaning that the user agent (or other external software components) are responsible for managing the channel list and handling how channel lists from different sources are merged and ordered.

> Finally, would we need the `isSourceAvailable` method? I imagine the main use case for this is to check whether a subsequent call to `getSource` will succeed, which we can handle through the returned promise. Are there other use cases I've missed?

I'm not sure - I need to think about this a little more, if people think that this is a good way forward.

Steve.

**Comment by tidoust** [comment]
Trying to capture main feedback received from the Media Capture Task Force (additional notes are from me):

1. General agreement that attaching to `enumerateDevices` seems a bad idea
2. Attaching to `mediaDevices` is still doable. I note that having a dedicated `TVManager` interface allows us to add other methods and properties at that level, which we seem to need. If we end up with an empty interface for some reason, we can always decide to re-attach to `mediaDevices` later on.
3. General agreement that using `MediaStreamTrack` constraints to switch to another channel seems a bad idea.
4. Streams are lightweight and are just a control surface. We could consider revisiting our decision to reuse streams across channels. I suspect the main arguments in favor/against reuse will be linked to locking the underlying resources.
5. Do we need MediaStreams at all? If we don't really plan to use constraints at the track level, we could just as well adopt a design similar to Media Source Extensions and attach to media elements directly.
6. Adding a `stop` method at the stream level seems a bad idea
7. `isRecordable` could perhaps be implemented by assigning a random identity at the streams level (?)

**Comment by tidoust** [comment]
@stevem-tw 

> The TVSource acts as a proxy for the various hardware and software resources needed to acquire a media stream, although the details of these resources are hidden from the app

Just to make sure I got that right. A `TVSource` in this model creates at most one stream. An app won't be able to tune to multiple channels at once using the same `TVSource`, even if it calls `tuneToChannel` multiple times. It needs two `TVSource` instances for that to be possible. Is that correct?

When would the user be prompted for permission in regular web applications (type 3)? As a result of a call to `getSource`? That seems needed not to leak information about available sources to arbitrary origins.

Delivery type makes sense as a criterion for users to grant permission ("this app wants to access your FM radio, do you agree?"). Quality level does not make a lot of sense ("this app wants to access your HD channels, do you agree?"). We may want to mandate the delivery type constraint to avoid asking the user too generic a question ("this app wants to access your broadcast signals, do you agree?"). Once user grants permission for an app to access a specific delivery type, permission can probably be assumed for further calls to `getSource` (and `isSourceAvailable`) for that same delivery type.

Also, the user would not be asked for permission to access a particular channel if it gets prompted after the call to `getSource`, unless we ask her again when calling to `tuneToChannel`. Or do you envision that the channel name could be one of the constraints that could be provided to `getSource`? This may be an edge case, I do not know whether apps will only want to tune in to one particular channel.

**Comment by stevem-tw** [comment]
@tidoust 

> Just to make sure I got that right. A TVSource in this model creates at most one stream. An app won't be able to tune to multiple channels at once using the same TVSource, even if it calls tuneToChannel multiple times. It needs two TVSource instances for that to be possible. Is that correct?

Yes

> When would the user be prompted for permission in regular web applications (type 3)? As a result of a call to getSource? That seems needed not to leak information about available sources to arbitrary origins.

Good question :)  I havent thought about this much, but my first reaction is that a call to getSource() is what should trigger prompting the user for permission.  It doesn't leak information on its own, but it's probably a the best gatekeeper to privileged functionality (tuning, getting a channel list, etc.) that lets us minimize the number of prompts to the user.

> Delivery type makes sense as a criterion for users to grant permission ("this app wants to access your FM radio, do you agree?"). Quality level does not make a lot of sense ("this app wants to access your HD channels, do you agree?"). We may want to mandate the delivery type constraint to avoid asking the user too generic a question ("this app wants to access your broadcast signals, do you agree?"). Once user grants permission for an app to access a specific delivery type, permission can probably be assumed for further calls to getSource (and isSourceAvailable) for that same delivery type.

Yes, but mandating the delivery type means that an application *has* to care about this, when it otherwise may not.  We may want to say that if a delivery type isn't specified then the user agent may prompt to give access to all delivery types, but I'm not sure if that's a good idea or not - I havent thought about it enough.

> Also, the user would not be asked for permission to access a particular channel if it gets prompted after the call to getSource, unless we ask her again when calling to tuneToChannel. Or do you envision that the channel name could be one of the constraints that could be provided to getSource? This may be an edge case, I do not know whether apps will only want to tune in to one particular channel.

I thought about supplying the channel name as one of the constraints, but that then gets complicated (e.g. if you know the tuning parameters, can you tune directly to that channel?).  As we've mentioned elsewhere, you could use constraints as a way of tuning - to change channel, you apply a new set of constraints - but that is something that needs more thought and still seems a little unnatural to me (although with this model, it's becoming less unnatural...)

**Comment by stevem-tw** [comment]
Following some discussion on this, the following additional points were raised:

The first issue that was discussed was around the constraints that can be applied when getting a source. Obvious ones are source type, video resolution, tuning step size (for radio), and possibly channel (so that an app can say "give me a source that can play this channel").  There may be others, but this needs more thought.

Another issue that was discussed was around permissions, and at what stage the user should grant permission for an app to be able to access the functionality, and how this should be presented to the user.  This discussion covered several topics that affected the API, including:

1. Should an app need user permission to access the channel list? Conclusion: yes, because the channel list provides further information that can be used for fingerprinting a device.
2. Should separate permissions be needed for "read-only" access to the API (i.e. read the channel list and present the current channel) vs. "read-write access (i.e. changing the current channel or starting to present a new channel)?  Conclusion: no, because we couldn't think of use cases where the distinction mattered.
3. What permission should the user actually be asked for, such that they will understand the question and give informed consent?  Conclusion: the user should be asked whether they want to allow access to broadcast TV or broadcast radio functionality, rather than anything narrower which could be too intrusive.

As a result of this discussion, there was a proposal that the API have two separate high-level entry points: one for TV functionality and one for radio functionality.  This could be two methods on the Window or Navigator object, 
```
Promise<TVManager> getTVManager()
Promise<TVManager> getRadioManager()
```

Note that these are separated by tuner type, in effect: radio channels delivered by DVB-T would still be accessed through getTVManager() rather than through getRadioManager().  

This has two purposes: first, to make it possible to ask for user permission in the way described above, and second to separate out the functionality where it makes sense to do so.  For instance, in a device that has both an FM tuner and a DVB-T tuner, providing a unified channel list that covers both tuners is probably meaningless; similarly, an FM or DAB radio device may not support recording, where a DVB-T or DVB-S device may do.  

The basic APIs would be the same in both cases, although there may be extra APIs which are only applicable in one of the two cases.

Note also that the name "TVManager" is still used in the example above, simply because we couldn't think of anything better.  Alternative suggestions are welcome!

**Comment by chrisn** [comment]
Thanks Steve. I'll try to summarise the API. Feel free to edit this post if it's not right.

```webidl
partial interface Navigator {
  // These methods may require the UA to request user permission
  Promise<TVManager> getTVManager();
  Promise<TVManager> getRadioManager();     
};
```

Alternatively, if the recommended way to extend Navigator is to add an interface (for now I've called it `TVControl`):

```webidl
partial interface Navigator {
  TVControl tv;
};

interface TVControl {
  Promise<TVManager> getTVManager();
  Promise<TVManager> getRadioManager();     
};

interface TVManager {
  TVSourceCapabilities getCapabilities();
  Promise<TVSource> getSource(TVSourceConstraints constraints);
  isSourceAvailable(TVSourceConstraints constraints);

  // A single channel list is managed by the UA.
  Promise<sequence<TVChannel>> getChannelList();

  // The scanning API is TBD (see issue #7):
  Promise<void> startScanning(optional TVStartScanningOptions options);
  Promise<void> stopScanning();

  readonly attribute boolean isScanning;
           attribute EventHandler onscanningstatechanged;

  // I've left out the recording, parental control, and CI card APIs
  // for brevity.
};

interface TVSource {
  TVSourceConstraints getConstraints();
  Promise<void> tuneToChannel(TVChannel channel);
  Promise<void> tuneTo(TVTuningParameters parameters);

  // See issue #29 -- alternatively: getDataComponents();
  Promise<TVApplication> getApplications();

  readonly attribute TVMediaStream? stream;

           attribute EventHandler oncurrentchannelchanged;
           attribute EventHandler oneitbroadcasted;
           attribute EventHandler onemergencyalerted;
};
```

The `TVChannel` and `TVProgram` interfaces remain largely as before.

**Comment by stevem-tw** [comment]
Chris, your API is pretty much exactly what I had in mind.

**Comment by tidoust** [comment]
One minor comment: I would merge `getXManager` methods into one that takes some sort of "type" parameter, e.g.:

```webidl
interface TVControl {
  Promise<TVManager> getManager(TVManagerType type);
}
```

I'm not exactly sure why, though. It just _seems_ cleaner and more extensible to me.

**Comment by chrisn** [comment]
Is there any feedback from other group members on this proposal? The next step will be to update the specification and work on the detail of the updated API, but I want to ask for consensus among WG members before doing that.

**Comment by chrisn** [comment]
I received this feedback from Mr Igarashi:

> My concern was that the implementations of sources could be independent so that the capabilities of sources may be constrained.
>
> Looking over the change, I do not prefer that TV Managers handles all channels for all sources. Ideally, an application might want to seamlessly handle channels across sources, e.g. DVB-T, DVB-S, IPTV services, OTT services, HDMI, etc. However, the idea can not be supported with some existing systems where each source is implemented independently or it is not allowed to handle sources seamlessly due to some business issues. Because of those reasons, the most UI on TV systems is that users have to select sources and then select a channel of the source. To support such existing systems, I think that the TV Manager should return list of sources and then a source provides a list of channels.
>
> In terms of constrains among sources, I need further study. One question is if TV source can represents only constrains among sources or it can present constrains among sources as well. I know a TV system which supports PinP screens which slow only two video from two DVB-T tuners and two DVB-S tuner. Since constrains on sources are very completed and implementation specific, I do not think that the TV Control WG can define the best mechanize to cover all cases of constrains. It would be OK even if the capability and constrains can not be pre-deterministic, i.e. try-and-error would be OK.
>
> My suggestion is like the following which is very similar to the original model of TV Control API. The Mozilla's model would match the existing TV systems. I wonder that the proposal idea matches subsidized STBs.

```webidl
partial interface Navigator {
  TVControl tv;
}

interface TVControl {
  Promise<TVManager> getTVManager();
  Promise<TVManager> getRadioManager();
}

interface TVManager {
  TVSourceCapabilities getCapabilities();
  Promise<TVSource> getSource(TVSourceConstraints constraints);
};

interface TVSource {
  isSourceAvailable(TVSourceConstraints constraints);
  TVSourceConstraints getConstraints();
  // A single channel list is managed by the UA.
  Promise<sequence<TVChannel>> getChannelList();

  // The scanning API is TBD (see issue #7):
  Promise<void> startScanning(optional TVStartScanningOptions options);
  Promise<void> stopScanning();

  readonly attribute boolean isScanning;
           attribute EventHandler onscanningstatechanged;

  // I've left out the recording, parental control, and CI card APIs
  // for brevity.

  Promise<void> tuneToChannel(TVChannel channel);
  Promise<void> tuneTo(TVTuningParameters parameters);

  // See issue #29 -- alternatively: getDataComponents();
  Promise<TVApplication> getApplications();

  readonly attribute TVMediaStream? stream;

           attribute EventHandler oncurrentchannelchanged;
           attribute EventHandler oneitbroadcasted;
           attribute EventHandler onemergencyalerted;
};
```

**Comment by tidoust** [comment]
That's interesting. I do not have particular insights on business issues. Would a "flat" merge of all channels, essentially the concatenation of all channels per source without any attempt to merge duplicates, be forbidden as well?

I do not see any good way to have a model where the app can both retrieve a complete list of channels when possible and retrieve a list of channels per source if not. I mean, we could also keep a `getChannelList` method at the `TVManager` level, which would fail if the implementation cannot merge lists of channels, but that would force applications to deal with two levels to retrieve channels, and that seems wrong.

If `getChannelList` sticks at the `TVSource` level, I note we can get rid of the `TVControl` level, as user permission could then be acquired when `getSource` is called.

**Comment by stevem-tw** [comment]
I understand Igarashi-san's point, and there are differences between the models used for retail TVs and for subsidized STBs.

My view on the issue of constraints is that constraints should work across all sources. The case where a device may only be able to support a certain number of concurrent streams, or may have limited ability to support multiple source types at the same time, for whatever reason, is one that needs to be supported.  There will be differences between platforms that mean these capabilities and constraints will be different, but I think that the constraints mechanism is flexible enough to deal with this.

For me, the problem with having the channel lists on the sources is this: if the device UI manages the channel list, and has an option to show the user a merged channel list, an app can't create a merged channel list that matches the one the device UI will show (because the user may have re-ordered the channel list in the device UI, for instance).  Unless there is an API for getting the channel list on the TVManager object, I don't see a way to solve this.

I see a couple of possible solutions for dealing with the case where sources can't be handled seamlessly - whether for technical or business reasons.  

One option is to simply say that when the user selects which source they want, this is done *outside* the scope of the TV control API.  An app will then see only those sources which the user has currently selected: for instance, if the user has selected the DVB-S source, then in the set of capabilities an app will only see DVB-S as a supported source type.  Similarly, when the app calls TVManager.getChannelList(), it only gets a list of the channels available over DVB-S.

The second approach is for TVManager.getChannelList() to take a set of constraints as a parameter, where those constraints specify which source (or set of sources) should be included in the returned channel list.  An implementation can then return an appropriate response if the app requests channels from a set of sources that can't be handled seamlessly. 

Both of these approaches let us solve the problem without needing getChannelList() methods in different places to handle the two use cases: the question then becomes, in the case where sources can;'t be handled seamlessly, does the user agent need to be aware of this, or is that selection outside the scope of the user agent?

Steve.

**Comment by chrisn** [comment]
I think I'd prefer the option where the source selection is in-scope of the API, and to investigate the capabilities and constraints needed as a next step. Do other people have a view?

**Comment by stevem-tw** [comment]
I had a new draft of the spec prepared where this was the case, so I've committed that.  Comments on the new draft are welcome.

**Comment by tidoust** [comment]
Thanks for the update, @stevem-tw. The new API follows the IDL we've discussed at length in this issue, so that looks good to me :)

I note a few inconsistencies in the spec below. Some of them can be fixed right away. Some of them could be addressed in separate issues later on:

- Given that the notion of tuner disappears from this redesign, I would drop the `TVSource.tuner` property and the `TVTuner` interface right away, otherwise the spec is somewhat confusing. The `stream` property should move to `TVSource`.
- Similarly, the `tunerchanged` event should be dropped. This begs the question as to what happens when a tuner is added/removed. Most probably the list of channels changes, which would trigger a `channellistchange` event (which does not exist yet). Or capabilities change and we need to create a `capabilitieschange` event, but that sounds a bit overkill: when are tuners added/removed? E.g. does that really happen without restarting the device?
- I see you put `<dfn>` within IDL blocks, for instance for `TVSourceCapabilities`, `TVSourceConstraints`.. This confuses ReSpec and breaks links to the underlying dictionary and interface. The `<dfn>` should rather be with the wrapping header, as in: `<h2><dfn>TVSourceConstraints</dfn>` Interface</h2>`.
- In `TVSourceCapabilities`, `deliverySystem` should be of type `sequence<TVSourceType>` to target the enumeration, not `sequence<DOMString>`.
- In `TVSourceConstraints`, `deliverySystem` is defined as `boolean`. I think it should of type `TVSourceType` (or perhaps `sequence<TVSourceType>` if we can think of use cases where applications will want to request more than source type at a time.
- In `TVSource`, what does `getSettings` do? It seems a duplicate of `getConstraints`. I would drop it and drop `TVSourceSettings` as well.
- In `TVControl`, I would rather have one `getManager` method that takes a parameter of type `TVManagerType`, which would be a short enum with "tv" and "radio". I believe it is cleaner than having two separate methods.

**Comment by paulhiggs** [comment]
We probably need to keep some notion of "tuner" in the specification, even if from an application or API level we are not binding channels directly to it.
The "dynamic tuner model" was added primarily to support USB type tuner devices (which is something we need to continue to support). When a tuner is added, a channel scan may need to be performed depending on when that tuner was last used. 
Consider a STB with some built in tuner (s). The user would have initiated a channel scan at some times and have a channel list in NV memory. If a USB tuner that has never been used is connected, then there would be no change to the channel list. A scan could be initiated through a native menu/application, but that beaks the uniformity and consistency we look to solve.

I am not sure how other W3 specs handle dynamic hardware, such as the attachment and removal of peripherals. Perhaps we need some peripheral category (i.e. "broadcast-tuner") that triggers some capability change notification in our API.

**Comment by stevem-tw** [comment]
I was assuming that this would be reflected through a change in the capabilities, but you're right that there may need to be some signalling that the capabilities have changed.  [Section 9.2 of the media capture and streams spec](https://www.w3.org/TR/mediacapture-streams/#mediadevices) shows how they handle this case, but I'm not sure we'd want to be able to enumerate the devices/tuners in the way that they support.
